kernel/kernel-api/src/test/scala/io/delta/kernel/internal/util/SchemaUtilsSuite.scala

```diff
@@ -310,6 +311,140 @@ class SchemaUtilsSuite extends AnyFunSuite {
+
+    val schemaChanges = SchemaUtils.computeSchemaChangesById(fieldMappingBefore, fieldMappingAfter)
+
+    assert(schemaChanges.addedFields().isEmpty)
+    assert(schemaChanges.removedFields().isEmpty)
+    assert(schemaChanges.updatedFields().size() == 1)
+    assert(schemaChanges.updatedFields().get(0) ==
+      new Tuple2(fieldMappingBefore.get(1L), fieldMappingAfter.get(1)))
+  }
+
+  test("Compute schema changes with type changed columns") {
+    val fieldMappingBefore = Map(
+      1L -> new StructField("id", IntegerType.INTEGER, true)).map { case (k, v) =>
+      java.lang.Long.valueOf(k) -> v
+    }.asJava
+
+    val fieldMappingAfter = Map(
+      1L -> new StructField("promoted_to_long", LongType.LONG, true)).map { case (k, v) =>
+      java.lang.Long.valueOf(k) -> v
+    }.asJava
+
+    val schemaChanges = SchemaUtils.computeSchemaChangesById(fieldMappingBefore, fieldMappingAfter)
+
+    assert(schemaChanges.addedFields().isEmpty)
+    assert(schemaChanges.removedFields().isEmpty)
+    assert(schemaChanges.updatedFields().size() == 1)
+    assert(schemaChanges.updatedFields().get(0) ==
+      new Tuple2(fieldMappingBefore.get(1L), fieldMappingAfter.get(1)))
+  }
+
+  test("Compute schema changes with dropped fields") {
+    val fieldMappingBefore = Map(
+      1L -> new StructField("id", IntegerType.INTEGER, true),
+      2L -> new StructField("data", StringType.STRING, true)).map { case (k, v) =>
+      java.lang.Long.valueOf(k) -> v
+    }.asJava
+
+    val fieldMappingAfter = Map(
+      2L -> new StructField("data", StringType.STRING, true)).map { case (k, v) =>
+      java.lang.Long.valueOf(k) -> v
+    }.asJava
+
+    val schemaChanges = SchemaUtils.computeSchemaChangesById(fieldMappingBefore, fieldMappingAfter)
+
+    assert(schemaChanges.addedFields().isEmpty)
+    assert(schemaChanges.updatedFields().isEmpty)
+    assert(schemaChanges.removedFields().size() == 1)
+    assert(schemaChanges.removedFields().get(0) == fieldMappingBefore.get(1))
+  }
+
+  test("Compute schema changes with nullability change") {
+    val fieldMappingBefore = Map(
+      1L -> new StructField("id", IntegerType.INTEGER, true),
+      2L -> new StructField("data", StringType.STRING, true)).map { case (k, v) =>
+      java.lang.Long.valueOf(k) -> v
+    }.asJava
+
+    val fieldMappingAfter = Map(
+      1L -> new StructField("id", IntegerType.INTEGER, true),
+      2L -> new StructField("required_data", StringType.STRING, false)).map { case (k, v) =>
+      java.lang.Long.valueOf(k) -> v
+    }.asJava
+
+    val schemaChanges = SchemaUtils.computeSchemaChangesById(fieldMappingBefore, fieldMappingAfter)
+
+    assert(schemaChanges.addedFields().isEmpty)
+    assert(schemaChanges.removedFields().isEmpty)
+    assert(schemaChanges.updatedFields().size() == 1)
+    assert(schemaChanges.updatedFields().get(0) == new Tuple2(
+      fieldMappingBefore.get(2),
+      fieldMappingAfter.get(2)))
+  }
+
+  test("Compute schema changes with moved fields") {
```

Comment:
Ah, so by "moved fields" do you mean changes in a nested data type? This is an interesting case that I wonder how will play out in the real scenario. Since I think both nested fields will also have their own id -> struct field mapping.

I think maybe we should make this test case before/after map reflect that real case scenario.

Reply:
> Ah, so by "moved fields" do you mean changes in a nested data type?
Yeah basically reordering fields within a struct

I can update the mapping though in the tests to include nested ID paths but to be clear the update is at the level of the struct containing the moved fields.

So for example in this case, it really is field with ID 1 (the struct) that should be considered the updated field because id/data moved around. Let's say `id` had ID 2, and `name` had ID 3, we would not expect schemaChanges for field with ID 2 nor field with 3.

Reply:
Hmm but also won't we see this field show up if there's any change to the nested fields in a struct? I.e. if i I change a comment in the column metadata of one of the nested columns that will also count right?

Reply:
> I can update the mapping though in the tests to include nested ID paths but to be clear the update is at the level of the struct containing the moved fields.

Yep I got this :)

Reply:
>Hmm but also won't we see this field show up if there's any change to the nested fields in a struct? I.e. if i I change a comment in the column metadata of one of the nested columns that will also count right?


Yeah that's correct. Any change or change(s) will be represented as a general update. If the concern is that this surfaces as multiple updates, I think that's actually OK for now because the validation we are going to do won't be impacted by this. It's also why I left the class package private because we are free to change the semantic of updated fields as we choose.